### PR TITLE
Issue 18 - Support conversion from any format to .g6 format

### LINKED ISSUE #18: Story in Epic 10: Support conversion from supported input formats to .g6 format
Part of Epic #10 : Support the following command-line option configuration: `planarity -t [-q] -tg I O`.

### PR DESCRIPTION
**Pull request by wbkboyer**
Resolves #18 

## Added
* `c/samples/`
  * <details><summary>nauty_example.g6.0-based.AdjList.out.txt.0-based.G6.out.g6</summary>Necessary for "round-trip" test in `runSpecificGraphTests`, where the input of the invocation of `runGraphTransformationTest` is `nauty_example.g6.0-based.AdjList.out.txt` (which itself is the adjacency list output resulting from running `planarity -t -ta` on `nauty_example.g6`)</details>
  * <details><summary>K10.g6.0-based.AdjList.out.txt.0-based.G6.out.g6</summary>Necessary for "round-trip" test in `runSpecificGraphTests`, where the input of the invocation of `runGraphTransformationTest` is `K10.g6.0-based.AdjList.out.txt` (which itself is the adjacency list output resulting from running `planarity -t -ta` on `K10.g6`)</details>
* `c/g6-write-iterator.h/.c` - "methods" operating over `G6WriteIterator` to handle encoding graphs in the `.g6` format and writing to string (i.e. the flow implemented in `_WriteGraphToG6String`, called by `gp_WriteToString`) or to file (i.e. the flow implemented in `_WriteGraphToG6FilePointer`, called by `gp_Write`)

## Updated
* `.vscode/`
  * `tasks.json`
    * Added build configuration for `gcc` on Linux (tested in Debian VM)
    * Updated `cwd` key for Windows MinGW `gcc` compiler build task to use `${workspaceFolder}` variable, and updated name of task
  * `launch.json`
    * Updated Windows launch configuration name as well as `preLaunchTask` value to correspond to Windows MinGW `gcc` build task (see above)
    * Added Linux `gcc` launch configuration with `preLaunchTask` set to the corresponding Linux `gcc` build task (see above) 

<details>
<summary>Minor updates</summary>

* `c/`
  * `samples/nauty_example.g6` - added `>>graph6<<` header (which is what will be output when by `G6WriteIterator` process, although it's not required by the file format specification; see [".g6" file format](https://users.cecs.anu.edu.au/~bdm/data/formats.txt))
  * `planarity.c`
    * `helpMessage` - updated command sample for `-t` flag to include `-t(gam)`
  * `planarityCommandLine.c`
    * `runSpecificGraphTests` - added calls to `runGraphTransformationTest` with command `-tg` for `K10.g6.0-based.AdjList.out.txt` and `nauty_example.g6.0-based.AdjList.out.txt` (with `inputInMemFlag` set to `TRUE` for both)
    * `runGraphTransformationTest` - updated error messages to indicate `-t(gam)` is expected. Also fixed memory leak (free `inputString` after we're done with it) and fixed `Result` return value to align with what `runSpecificGraphTests` expects.
    * `callTestGraphFunctionality` - updated docstring with `-t(gam)` and indicating `.g6`, adjacency list, and adjacency matrix transformation targets are allowed
  * `planarityTestGraphFunctionality.c`
    * `TestGraphFunctionality` - updated docstring to indicate `-t(gam)` expected; updated `if-else if-else` structure to incorporate `commandString` containing `g`, using new `MODE` as `WRITE_G6` (from `graph.h`)
  * `planarityUtils.c`
    * `GetTransformationName` - updated `switch` statement to include case `g` corresponding to `G6` component of output filename
  * `graphIO.c`
    * `gp_Write` - updated docstring to include mention of `WRITE_G6`, and added `switch` statement case for `Mode` to handle `WRITE_G6` (i.e. invoke `_WriteGraphToG6FilePointer` in `G6WriteIterator` to write `.g6` encoded graph _to a *file*_ ending in `.g6` extension)
    * `gp_WriteToString` - updated docstring to include mention of `WRITE_G6`, and added `switch` statement case for `Mode` to handle `WRITE_G6` (i.e. invoke `_WriteGraphToG6String` in `G6WriteIterator` to write `.g6` encoded graph _to a *string*_)
    * `_ReadAdjMatrix`, `_ReadAdjList`, `_ReadLEDAGraph` - changed calls to `gp_AddEdge` to `gp_DynamicAddEdge` so that we may read dense graphs
  * `strOrFile.h`
    * changed the names of members of `strOrFile` `struct` to `pFile`, `theStr`, and `theStrPos` so that this struct can be used when handling _writing_ `.g6` to file or string. Also, since the parameter list in the declaration includes names, I changed the names of the `FILE *` and `char *` parameter for `sf_New` initialization
    * Declared `sf_fputs` and `sf_getTheStr`
  * `strOrFile.c`
    * `sf_New` - changed parameter names to match the `strOrFile` definition in `strOrFile.h`; now we reference `pFile` (vs. `inputFile`) and `theStr` (vs. `inputStr`)
    * `sf_getc`, `sf_ungetc`, `sf_fgets` - changed dereferencing to match the `strOrFile` declaration in `strOrFile.h`
    * `sf_fputs` - called by `_beginG6WriteIteration` and by `_printEncodedGraph`;calls `fputs` if the `strOrFile` container contains a `FILE` pointer, or concatenates to `theStr` and updates `theStrPos` if the `strOrFile` container contains a string
  * `g6-read-iterator.h/.c` - changed dereferencing to match the `strOrFile` declaration in `strOrFile.h` 

</details>

### CODE REVIEW

File: .vscode/tasks.json [hunk 1]
```diff
@@ -65,6 +65,27 @@
             ],
             "group": "build",
             "detail": "Task generated by Debugger."
+        },
+        {
+            "type": "cppbuild",
+            "label": "C/C++: gcc build active file (Linux)",
```
Comment (wbkboyer) [inline .vscode/tasks.json:71]: Rename to "build project"

File: .vscode/tasks.json [hunk 2]
```diff
@@ -25,18 +25,18 @@
         },
         {
             "type": "cppbuild",
-            "label": "C/C++: gcc.exe build active file",
+            "label": "C/C++: gcc.exe build active file (Windows)",
             "command": "C:\\MinGW\\bin\\gcc.exe",
             "args": [
                 "-fdiagnostics-color=always",
-                "-DDEBUG", // Added so that _DEBUG is defined; see appconst.h
```
Comment (wbkboyer) [inline .vscode/tasks.json:32]: Restore these comments, as they were informative.

File: c/planarityCommandLine.c [hunk 3]
```diff
@@ -357,6 +366,23 @@ int runSpecificGraphTests(char *samplesDir)
 		Message("Transforming K10.g6 using file pointer to adjacency matrix failed.\n");
 	}
 
+	//	TRANSFORM TO .G6
+
+	// runGraphTransformationTest by reading from file
+	if (runGraphTransformationTest("-tg", "nauty_example.g6.0-based.AdjList.out.txt", TRUE) < 0)
```
Comment (john-boyer-phd) [inline c/planarityCommandLine.c:372]: Great to smoke test g6 writing.

File: c/planarityCommandLine.c [hunk 4]
```diff
@@ -538,7 +562,13 @@ int runGraphTransformationTest(char *command, char *infileName, int inputInMemFl
 
 	Message("\n");
 
-	return Result;
+	if (inputString != NULL)
```
Comment (john-boyer-phd) [inline c/planarityCommandLine.c:565]: Good catch.

File: c/strOrFile.c [hunk 5]
```diff
@@ -137,28 +137,73 @@ int numCharsRead = -1;
...
+
+/********************************************************************
+ sf_fputs()
+ Order of parameters matches stdio fputs().
+
+ First param is the string to append, and the second param is the
+ string-or-file container to which we wish to append.
+
+ On success, returns the number of characters written.
+ On failure, returns EOF. TODO: should we set error code?
```
Comment (john-boyer-phd) [inline c/strOrFile.c:156]: Remove TODO. Returning EOF is the error indicator for fputs, so it is sufficient for this too.

**Review by john-boyer-phd — CHANGES_REQUESTED** [review]
Excellent and good to go except tweaks to a few comments.

**Review by john-boyer-phd — APPROVED** [review]
Looks great!

**Pull request merged**